src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -567,13 +568,44 @@ private function writeProperty(&$object, $property, $value)
 
             $object->$property = $value;
         } elseif (self::ACCESS_TYPE_MAGIC === $access[self::ACCESS_TYPE]) {
-            $object->{$access[self::ACCESS_NAME]}($value);
+            $this->callMethod($object, $access[self::ACCESS_NAME], $value);
         } else {
             throw new NoSuchPropertyException($access[self::ACCESS_NAME]);
         }
     }
 
     /**
+     * Call a method and convert {@see \TypeError} to {@see InvalidArgumentException}.
+     *
+     * @param object $object
+     * @param string $method
+     * @param mixed  $value
+     */
+    private function callMethod($object, $method, $value) {
+        if (class_exists('TypeError')) {
+            // PHP 7
+            try {
+                $object->{$method}($value);
+            } catch (\TypeError $e) {
+                throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+            }
+
+            return;
+        }
+
+        // PHP 5
+        set_error_handler(function ($errno, $errstr) {
+            throw new InvalidArgumentException($errstr);
+        });
+
+        try {
+            $object->{$method}($value);
+        } finally {
```

Comment:
This may cause issues with PHP < 5.6 (a.k.a 5.5), which have a bugged implementation of `finally`.